agent/agent.go

```diff
@@ -1999,25 +2135,80 @@ func (a *Agent) AddServiceAndReplaceChecks(service *structs.NodeService, chkType
 func (a *Agent) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
 	a.stateLock.Lock()
 	defer a.stateLock.Unlock()
-	return a.addServiceLocked(service, chkTypes, persist, token, false, source)
+	return a.addServiceLocked(service, chkTypes, nil, true, persist, true, token, false, source)
 }
 
 // addServiceLocked adds a service entry to the service manager if enabled, or directly
 // to the local state if it is not. This function assumes the state lock is already held.
-func (a *Agent) addServiceLocked(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, replaceExistingChecks bool, source configSource) error {
+func (a *Agent) addServiceLocked(
+	service *structs.NodeService,
+	chkTypes []*structs.CheckType,
+	previousDefaults *structs.ServiceConfigResponse,
+	waitForCentralConfig bool,
+	persist bool,
+	persistServiceConfig bool,
+	token string,
+	replaceExistingChecks bool,
+	source configSource,
+) error {
 	if err := a.validateService(service, chkTypes); err != nil {
 		return err
 	}
 
 	if a.config.EnableCentralServiceConfig {
-		return a.serviceManager.AddService(service, chkTypes, persist, token, source)
-	}
+		return a.serviceManager.AddService(
+			service,
+			chkTypes,
+			previousDefaults,
+			waitForCentralConfig,
+			persist,
+			persistServiceConfig,
+			token,
+			replaceExistingChecks,
+			source,
+		)
+	}
+
+	// previousDefaults are ignored here because they are only relevant for central config.
+	return a.addServiceInternal(&addServiceRequest{
+		service:               service,
+		chkTypes:              chkTypes,
+		persistService:        nil,
+		persistDefaults:       nil,
+		persist:               persist,
+		persistServiceConfig:  false,
+		token:                 token,
+		replaceExistingChecks: replaceExistingChecks,
+		source:                source,
+	})
+}
 
-	return a.addServiceInternal(service, chkTypes, persist, token, replaceExistingChecks, source)
+type addServiceRequest struct {
+	service               *structs.NodeService
+	chkTypes              []*structs.CheckType
+	persistService        *structs.NodeService
+	persistDefaults       *structs.ServiceConfigResponse
+	persist               bool
+	persistServiceConfig  bool
+	token                 string
+	replaceExistingChecks bool
+	source                configSource
```

Comment:
I was going to comment on the number of arguments in `addServiceLocked` growing then saw this. Is there a reason we can't re-use this across the other `[aA]ddService(Locked)` methods? I don't feel super strongly just seems more consistent if we are bothering to define a struct in this case etc.

Reply:
The function signatures are not entirely the same. There is enough  overlap that I could just add the union of fields to the struct and just document with comments which fields are only meant for one or the other